create_primers.py

```diff
@@ -0,0 +1,261 @@
+    Primers are named as follows:
+
+    f"{prefix}-for-mut{i}{mutant}" for 5' tiling primers, where i is the site
+    mutagenized in the gene and mutant is the amino acid the site is mutated to
+    using that primer.
+
+    Returns a list of all these primers as *(name, sequence)* 2-tuples.
+    """
+    n = len(seq)
+    assert primerlength % 2 == 1, "primer length not odd"
+    initial_flanklength = (primerlength - 3) // 2
+    upperseq = ''.join([nt for nt in seq if nt.istitle()])
+    assert upperseq in seq, "upper case nucleotides not substring"
+    assert len(upperseq) % 3 == 0, "length of upper case not multiple of 3"
+    ncodons = len(upperseq) // 3
+
+    # Read in the mutations csv and make sure it is the right format
+    df = pd.read_csv(mutations_csv)
+    try:
+        sites = df['site'].tolist()
+    except:
+        print('mutations_csv must have column "site"')
+    try:
+        mutations = df['mutant'].tolist()
+    except:
+        print('mutations_csv must have column "mutant"')
+    mutations_dict = {}
+    for i, site in enumerate(sites):
+        if site not in mutations_dict:
+            mutations_dict[site] = [mutations[i]]
+        else:
+            mutations_dict[site].append(mutations[i])
+    # Make sure this is enough flanking sequence to make primers
+    initial_flanklength = (primerlength - 3) // 2
+    firstcodon = min(sites)
+    lastcodon = max(sites)
+    startupper = seq.index(upperseq) + ((firstcodon - 1) * 3)
+    if startupper < initial_flanklength:
+        raise ValueError("not enough 5' flanking nucleotides")
+    if n - len(upperseq) - seq.index(upperseq) + (len(upperseq) - (lastcodon - 1) * 3) < initial_flanklength:
+        raise ValueError("not enough 3' flanking nucleotides")
+    # Read in the codon frequency table, make dict for back translating
+    df = pd.read_csv('codon_frequencies.csv')
+    aas = df['aa'].tolist()
+    codons = df['codon'].tolist()
+    frequencies = df['frequency'].tolist()
+    back_t_dict = {}
+    for i, codon in enumerate(codons):
+        if aas[i] not in back_t_dict:
+            back_t_dict[aas[i]] = {}
+            back_t_dict[aas[i]][frequencies[i]] = codon
+        else:
+            back_t_dict[aas[i]][frequencies[i]] = codon
+    # Iterate through the codons and make the specific primers, with most
+    # frequent codon
+    primers = []
+    for icodon in range(ncodons):
+        # Determine what the insert should be for the codon
+        # Replace ambiguous_codon with codon insert later
+        codon_inserts = []
+        if icodon + 1 in mutations_dict:
+            for mutation in mutations_dict[icodon + 1]:
+                max_frequency = max(back_t_dict[mutation])
+                codon_insert = back_t_dict[mutation][max_frequency]
+                i = startupper + icodon * 3
+                primer = "%s%s%s" % (seq[i - initial_flanklength : i], codon_insert, seq[i + 3 : i + 3 + initial_flanklength])
```

Comment:
You use this in a lot of places, but I'm just commenting here. The "%s", "%d", etc. string formatting is a hold over from the earlier script, but it is also very Python 2. 

I'd recommend switching all your strings to f string formatting, which is much more readable in my opinion too. So the code would be:

`primer = f"{seq[i - initial_flanklength : i]}{codon_insert}{seq[i + 3 : i + 3 + initial_flanklength]}"`

Like in the line below.

I know there are lots of places where this change would need to be made, but I think it's worth making in order to be consistent and up-to-date.